src/libstd/sys/unix/process.rs

```diff
@@ -283,6 +363,24 @@ impl Process {
         if !setup(out_fd, libc::STDOUT_FILENO) { fail(&mut output) }
         if !setup(err_fd, libc::STDERR_FILENO) { fail(&mut output) }
 
+        let close_or_leak = |fd: RawFd| {
+            if cfg.leak_fds_whitelist.contains(&fd) {
+                let flag_orig = libc::fcntl(fd, libc::F_GETFD, 0);
+                if flag_orig < 0 {
+                    return;
+                }
+                let flag_new = flag_orig & !libc::FD_CLOEXEC;
+                if flag_orig != flag_new {
+                    let _ = libc::fcntl(fd, libc::F_SETFD, flag_new);
+                }
+            } else {
+                let _ = libc::close(fd);
+            }
+        };
+        if !cfg.leak_fds {
+            walk_open_fd(close_or_leak);
```

Comment:
Unfortunately, the situation with standard libraries using `pthread_atfork` to correctly unlock locks has always been messy (see https://bugzilla.redhat.com/show_bug.cgi?id=906468, see https://stackoverflow.com/questions/29106483/glibc-malloc-deadlock-during-string-reserve, see https://code.google.com/p/gperftools/issues/detail?id=496, see https://code.google.com/p/android/issues/detail?id=178033, see https://bugzilla.mozilla.org/show_bug.cgi?id=680190). These were just a few of the bugs I was able to quickly search for. Anyway, being able to use `malloc` after a fork in a multi-threaded process isn't supported by standards and not well supported across platforms.

One horrible hack to support this might be to `exec` a helper binary (which can be the same file as a linked-in shared object).